### PR TITLE
`Lectures`: Add visible date to lectures

### PR DESCRIPTION
**Pull request by lennart-keller**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

Lectures are always visible to students, even if they are not relevant yet.

### Description

The lecture table now contains an additional column: visible_date
It can be set by users in the same way as the start and end dates through the client.
When students request the lectures from the server, they are filtered for visibility.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Course with messaging enabled

1. Login as an instructor
2. Create a new lecture with visible date
3. Confirm that the visible cannot be set after the start date
4. Import the newly created lecture and confirm that the visible date is the same as for the original lecture
5. Set the visible date to the future
6. Login as a student and confirm that the lecture is invisible
7. Confirm that the lecture channel is also invisible
8. Change the visible to the past and confirm that the lecture is visible to the student
9. Confirm that the lecture channel is visible to the student

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [ ] Test 2

### Screenshots

When creating a new lecture, there is a new input field for the visible date:
input field

The visible date value is shown in the overview after import/create/edit:
lecture overview

And it is shown in the overview list of lectures in an additional column:
lectures list

For students, lectures with a visible date in the future are hidden:
hidden lecture

As soon as the visible date passes, the lecture becomes visible:
visible lecture

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/LectureService.java [hunk 1]
```diff
@@ -73,9 +73,13 @@ public void filterActiveAttachmentUnits(Lecture lectureWithAttachmentUnits) {
      * @param user                    the user for which this call should filter
      * @return lectures with filtered attachments
      */
-    public Set<Lecture> filterActiveAttachments(Set<Lecture> lecturesWithAttachments, User user) {
+    public Set<Lecture> filterVisibleLecturesWithActiveAttachments(Set<Lecture> lecturesWithAttachments, User user) {
         Set<Lecture> lecturesWithFilteredAttachments = new HashSet<>();
         for (Lecture lecture : lecturesWithAttachments) {
+            if (authCheckService.isOnlyStudentInCourse(lecture.getCourse(), user) && !lecture.isVisibleToStudents()) {
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/LectureService.java:79]: Are all lectures here from the same course or can there be multiple courses? If it's only one course, consider moving the `isOnlyStudent` check out of the loop.

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java [hunk 2]
```diff
@@ -208,11 +208,14 @@ public ResponseEntity<Set<Lecture>> getLecturesWithSlidesForCourse(@PathVariable
     public ResponseEntity<Lecture> getLecture(@PathVariable Long lectureId) {
         log.debug("REST request to get lecture {}", lectureId);
         Lecture lecture = lectureRepository.findById(lectureId).orElseThrow();
+        authCheckService.checkHasAtLeastRoleForLectureElseThrow(Role.STUDENT, lecture, null);
+        authCheckService.checkIsAllowedToSeeLectureElseThrow(lecture, userRepository.getUserWithGroupsAndAuthorities());
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java:212]: This already checks for the student role so the call above is not needed.

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java [hunk 3]
```diff
@@ -268,6 +271,7 @@ public ResponseEntity<Lecture> getLectureWithDetails(@PathVariable Long lectureI
             return ResponseEntity.badRequest().build();
         }
         User user = userRepository.getUserWithGroupsAndAuthorities();
+        authCheckService.checkIsAllowedToSeeLectureElseThrow(lecture, user);
         authCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, user);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java:275]: same, student check no longer needed here.

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java [hunk 4]
```diff
@@ -289,6 +293,7 @@ public ResponseEntity<Lecture> getLectureWithDetailsAndSlides(@PathVariable Long
         if (course == null) {
             return ResponseEntity.badRequest().build();
         }
+        authCheckService.checkIsAllowedToSeeLectureElseThrow(lecture, userRepository.getUserWithGroupsAndAuthorities());
         authCheckService.checkHasAtLeastRoleInCourseElseThrow(Role.STUDENT, course, null);
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java:297]: same

File: src/test/java/de/tum/in/www1/artemis/service/LectureServiceTest.java [hunk 5]
```diff
@@ -66,6 +68,14 @@ void initTestCase() throws Exception {
         // always use the lecture and course with the smallest ID, otherwise tests below related to search might fail (in a flaky way)
         course = courseRepository.findByIdWithLecturesAndLectureUnitsElseThrow(courses.stream().min(Comparator.comparingLong(DomainObject::getId)).orElseThrow().getId());
         lecture = course.getLectures().stream().min(Comparator.comparing(Lecture::getId)).orElseThrow();
+        hiddenLecture = course.getLectures().stream().max(Comparator.comparing(Lecture::getId)).orElseThrow();
```
Comment (Strohgelaender) [inline src/test/java/de/tum/in/www1/artemis/service/LectureServiceTest.java:71]: lecture and hidden lecture now refer to the same object, this does not seem right.

Comment (lennart-keller) [reply src/test/java/de/tum/in/www1/artemis/service/LectureServiceTest.java:71]: They do not (I am using min and max here and there are two lectures). I didn't want to remove the min/max due to the comment regarding flakiness above. I added an assertion below to ensure that lecture and hiddenLecture are different.

File: src/main/java/de/tum/in/www1/artemis/service/LectureService.java (hunk 1 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/LectureService.java]: They are always from the same course. I moved the check outside of the loop.

File: src/test/java/de/tum/in/www1/artemis/service/LectureServiceTest.java (hunk 5 above)
Comment (Strohgelaender) [reply src/test/java/de/tum/in/www1/artemis/service/LectureServiceTest.java]: oh, did not see that you used min and max.

Can you move this into a local variable of the test case? This lecture is currently only needed in this one test and not somewhere else.

Comment (lennart-keller) [reply src/test/java/de/tum/in/www1/artemis/service/LectureServiceTest.java]: Already done ;)

**Review by laadvo — APPROVED** [review]
Tested on ts1, the lecture visible date is working as expected

**Review by JohannesStoehr — APPROVED** [review]
Code changes look good to me

File: src/main/java/de/tum/in/www1/artemis/service/LectureService.java [hunk 6]
```diff
@@ -76,16 +76,23 @@ public void filterActiveAttachmentUnits(Lecture lectureWithAttachmentUnits) {
...
+     * Filter active attachments for a set of lectures. All lectures must be from the same course.
      *
+     * @param course                  course all the lectures are from
      * @param lecturesWithAttachments lectures that have attachments
      * @param user                    the user for which this call should filter
      * @return lectures with filtered attachments
      */
-    public Set<Lecture> filterActiveAttachments(Set<Lecture> lecturesWithAttachments, User user) {
+    public Set<Lecture> filterVisibleLecturesWithActiveAttachments(Course course, Set<Lecture> lecturesWithAttachments, User user) {
+        if (!authCheckService.isOnlyStudentInCourse(course, user)) {
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/LectureService.java:87]: Can this be called with a user who is not part of the course?

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java [hunk 7]
```diff
@@ -306,8 +308,9 @@ public ResponseEntity<Lecture> getLectureWithDetailsAndSlides(@PathVariable Long
     @GetMapping("/lectures/{lectureId}/title")
     @EnforceAtLeastStudent
     public ResponseEntity<String> getLectureTitle(@PathVariable Long lectureId) {
-        final var title = lectureRepository.getLectureTitle(lectureId);
-        return title == null ? ResponseEntity.notFound().build() : ResponseEntity.ok(title);
+        Lecture lecture = lectureRepository.findByIdElseThrow(lectureId);
+        authCheckService.checkIsAllowedToSeeLectureElseThrow(lecture, userRepository.getUserWithGroupsAndAuthorities());
+        return ResponseEntity.ok(lecture.getTitle());
     }
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java:314]: Can you doublecheck the usage of this endpoint? I think for some reason we decided to not do access checks when retriving only the title.

File: src/main/java/de/tum/in/www1/artemis/service/LectureService.java (hunk 6 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/service/LectureService.java]: I check for `isAtLeastTeachingAssistantInCourse` now.

File: src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java (hunk 7 above)
Comment (lennart-keller) [reply src/main/java/de/tum/in/www1/artemis/web/rest/LectureResource.java]: You are right, I reverted the changes I made to that endpoint.

**Review by laadvo — APPROVED** [review]
re-tested on ts1, the visible date of lectures is working as expected

**Review by krusche — APPROVED** [review]
Code looks good

**Pull request merged**